libmamba/src/core/util.cpp

```diff
@@ -1060,72 +997,214 @@ namespace mamba
             {
-                LOG_TRACE << "Removing dangling lock file '" << m_lock << "'";
-                m_lockfile_existed = false;
+                const auto absolute_file_path = fs::absolute(file_path);
+                std::scoped_lock lock{ mutex };
+
+                const auto it = locked_files.find(absolute_file_path);
+                if (it != locked_files.end())
+                {
+                    if (auto lockedfile = it->second.lock())
+                    {
+                        // TODO: just `return lockedfile;` once we decide to remove the exception
+                        throw_duplicate_lockfile_in_process(absolute_file_path, noop);
+                    }
+                }
+
+                // At this point, we didn't find a lockfile alive, so we create one.
+                auto lockedfile = std::make_shared<LockFileOwner>(absolute_file_path, timeout);
+                auto tracker = std::weak_ptr{ lockedfile };
+                locked_files.insert_or_assign(absolute_file_path, std::move(tracker));
+                fd_to_locked_path.insert_or_assign(lockedfile->fd(), absolute_file_path);
+                assert(is_lockfile_locked(*lockedfile));
+                return lockedfile;
             }
-#endif
+
+            // note: the resulting value will be obsolete before returning.
+            bool is_locked(const fs::u8path& file_path) const
+            {
+                const auto absolute_file_path = fs::absolute(file_path);
+                std::scoped_lock lock{ mutex };
+                auto it = locked_files.find(file_path);
+                if (it != locked_files.end())
+                    return !it->second.expired();
+                else
+                    return false;
+            }
+
+            // note: the resulting value will be obsolete before returning.
+            bool is_locked(int fd) const
+            {
+                std::scoped_lock lock{ mutex };
+                const auto it = fd_to_locked_path.find(fd);
+                if (it != fd_to_locked_path.end())
+                    return is_locked(it->second);
+                else
+                    return false;
+            }
+
+        private:
+            // TODO: replace by something like boost::multiindex or equivalent to avoid having to
+            // handle 2 hashmaps
+            std::unordered_map<fs::u8path, std::weak_ptr<LockFileOwner>>
+                locked_files;  // TODO: consider replacing by real concurrent set to avoid having to
+                               // lock the whole container
+
+            std::unordered_map<int, fs::u8path>
+                fd_to_locked_path;  // this is a workaround the usage of file descriptors on linux
+                                    // instead of paths
+            mutable std::recursive_mutex
+                mutex;  // TODO: replace by synchronized_value once available
+        };
+
+        static LockedFilesRegistry files_locked_by_this_process;
```

Comment:
Do we need to declare this variable static since it is defined in an anonymous namespace?

Reply:
You are correct that this is redundant. I added it in case it was not clear to everybody, but can remove it without change.